bin/testObservability/reporter/index.js

```diff
@@ -0,0 +1,584 @@
+        const {browser, platform} = this.platformDetailsMap[process.pid][test.title];
+        testData = {...testData, integrations: {
+          'local_grid': {
+            'capabilities': {},
+            'platform': os,
+            'platform_version': os_version,
+            'browser': browser.name,
+            'browser_version': browser.majorVersion
+          }
+        }};
+        if(eventType === "TestRunFinished" || eventType === "TestRunSkipped") {
+          delete this.platformDetailsMap[process.pid][test.title];
+        }
+      }
+
+      if (eventType === "TestRunSkipped" && !testData['started_at']) {
+        testData['started_at'] = testData['finished_at'];
+      }
+
+      try {
+        if(eventType.match(/HookRun/)) {
+          [testData.hook_type, testData.name] = getHookDetails(test.fullTitle() || test.originalTitle || test.title);
+          if(eventType === "HookRunFinished") {
+            if(testData.result === 'pending') testData.result = 'passed';
+            if(testData.hook_type == 'before each' && testData.result === 'failed' && ( !this.runStatusMarkedHash[test.ctx.currentTest.testAnalyticsId] )) {
+              if(test.ctx.currentTest.testAnalyticsId) this.runStatusMarkedHash[test.ctx.currentTest.testAnalyticsId] = true;
+              test.ctx.currentTest.state = STATE_FAILED;
+              await this.sendTestRunEvent(test.ctx.currentTest,undefined,true);
+            }
+          }
+        } else if(eventType.match(/TestRun/)) {
+          mapTestHooks(test);
+        }
+      } catch(e) {
+        debug(`Exception in processing hook data for event ${eventType} with error : ${e}`, true, e);
+      }
+
+      const failure_data = testData['failure'][0];
+      if (failure_data) {
+        testData['failure_backtrace'] = failure_data['backtrace']
+        testData['failure_reason_expanded'] = failure_data['expanded']
+      }
+
+      if(["TestRunFinished","TestRunSkipped"].includes(eventType)) {
+        testData.hooks = getHooksForTest(test);
+      }
+
+      let uploadData = {
+        event_type: eventType === "TestRunSkipped" ? "TestRunFinished" : eventType,
+      }
+
+      if(eventType.match(/HookRun/)) {
+        testData['hook_type'] = HOOK_TYPES_MAP[testData['hook_type']];
+        uploadData['hook_run'] = testData;
+      } else {
+        uploadData['test_run'] = testData;
+      }
+      
+      await uploadEventData(uploadData);
+
+      if(process.env.observability_framework_version || this.currentCypressVersion) {
+        const buildUpdateData = {
+          event_type: 'BuildUpdate',
+          'misc': {
+            observability_version: {
+              frameworkName: "Cypress",
+              sdkVersion: process.env.OBSERVABILITY_LAUNCH_SDK_VERSION,
+              frameworkVersion: ( process.env.observability_framework_version || this.currentCypressVersion )
+            }
+          }
+        };
+        await uploadEventData(buildUpdateData);
```

Comment:
Does this build update gets sent for every test run event? If so, can we send it only single time?

Reply:
Done. Sending once per reporter process now.